### PR TITLE
HPCC-17914 Optimize the performance for WildMatch

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Added unit tests for WildMatch - functionality and timing.

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman please review.  Also, can you think of any other test cases that I haven't included.  Some sample timing comparisons for the timing results:

Before:
10000: 53 ms
100000: 2389 ms
1000000: 227851 ms

After:
10000: 7 ms
100000: 380 ms
1000000: 26029 ms

**Comment by richardkchapman** [comment]
@jakesmith Please review

File: testing/unittests/jlibtests.cpp [hunk 1]
```diff
@@ -1009,4 +1011,167 @@ class JlibReaderWriterTestTiming : public CppUnit::TestFixture
...
+        std::unique_ptr<char[]> search(generateSearchString(length));
+        CCycleTimer timer;
+        testPatterns(search.get(), patterns);
+        if (reportTiming)
+            printf("%u: %u ms\n", length, timer.elapsedMs());
+    }
+    char * generateSearchString(size_t len)
+    {
+        const unsigned options = 10;
+        unsigned counts[options];
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: options and counts unused.

File: testing/unittests/jlibtests.cpp [hunk 2]
```diff
@@ -1009,4 +1011,167 @@ class JlibReaderWriterTestTiming : public CppUnit::TestFixture
...
+        char * target = new char[len+1];
+        fillSearchString(target, len);
+        target[len] = 0;
+        return target;
+    }
+
+    void fillSearchString(char * target, size_t len)
+    {
+        for (unsigned repeat =0; ; repeat++)
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: trivial: odd spacing before =0

File: testing/unittests/jlibtests.cpp [hunk 3]
```diff
@@ -1009,4 +1011,167 @@ class JlibReaderWriterTestTiming : public CppUnit::TestFixture
...
+            if (*pattern == '~')
+            {
+                nocase = true;
+                pattern++;
+            }
+            bool evaluated = WildMatch(search, pattern, nocase);
+            CPPUNIT_ASSERT_EQUAL_MESSAGE(pattern, expected, evaluated);
+        }
+    }
+
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: trivial: extra newline

**Comment by jakesmith** [comment]
@ghalliday - can't see any issues, test coverage looks good to me.
A couple of trivial comments.

**Comment by ghalliday** [comment]
@richardkchapman changes pushed

**Pull request merged**